src/cms/logic.py

```diff
@@ -0,0 +1,323 @@
+
+
+def gobble_sibling_text(sibling, original_part):
+    """
+    Recursively collects a series of parallel elements for
+    indexing together, stopping before another heading
+    """
+
+    # Handle no sibling
+    if not sibling:
+        return ''
+
+    # Handle another heading
+    if sibling.name == original_part.name:
+        return ''
+    if isinstance(sibling, Tag) and sibling.find(original_part.name):
+        return ''
+
+    # Recursively get sibling text
+    if isinstance(sibling, NavigableString):
+        sibling_text = sibling.strip()
+    else:
+        sibling_text = sibling.get_text().strip()
+    next_sibling_text = gobble_sibling_text(
+        sibling.next_sibling,
+        original_part
+    )
+
+    # Decompose
+    if isinstance(sibling, NavigableString):
+        sibling.replace_with('')
+    else:
+        sibling.decompose()
+
+    gobbled_text = sibling_text + ' ' + next_sibling_text
+    return gobbled_text.strip()
+
+
+def get_text_for_parent(parent, original_part):
+    if not parent:
+        return ''
+
+    text = gobble_sibling_text(parent.next_sibling, original_part)
+    if len(text.strip()) > len(original_part.get_text().strip()):
+        return text
+    else:
+        return get_text_for_parent(parent.parent, original_part)
+
+
+def get_text_for_header(part):
+    """
+    Gets the sibling text or parent text
+    when given an h2, h3, or h4
+    """
+
+    # Try gobbling first
+    text = gobble_sibling_text(part.next_sibling, part)
+    if len(text.strip()) > 0:
+        return text
+
+    # Otherwise go to parent
+    return get_text_for_parent(part.parent, part)
+
+
+def add_part_as_doc(part, part_url, part_name, part_text):
+    if not part_name or not part_text:
+        # HTML tree with id has no heading
+        # or no text content
+        return
+    add_search_index_document(part_url, part_name, part_text)
+    part.decompose()
+
+
+def add_searchable_page_parts(url, body):
+
+    headings = ['h4', 'h3', 'h2']
+
+    def get_part_url(part):
+        part_id = part.get('id', '')
+        if part_id:
+            return urljoin(url, '#' + part_id)
+        else:
+            return remove_fragment(url)
+
+    for h in headings:
+        for part in body.find_all(h, id=True):
+            if not part:
+                continue
+            part_url = get_part_url(part)
+            part_name = part.get_text().strip()
+            part_text = get_text_for_header(part)
+            add_part_as_doc(part, part_url, part_name, part_text)
+
+
+def get_page(url):
+    time.sleep(.1)
+    try:
+        fetched_urls.add(url)
+        headers = {
+            'Accept': 'text/html; charset=utf-8'
+        }
+        response = requests.get(url, headers=headers)
+    except requests.exceptions.ConnectionError:
+        logger.warn(f'Could not access {url}')
+        logger.warn('Please run server to index site search')
+        return
+
+    if response.status_code != 200:
+        return
+
```

Comment:
If feel like something should be logged here too in some circumstances: If we've determined that the URL should be indexable, it should let us know when a 500 or 404 is returned. I imagine you are also guarding against 401 codes..